### PR TITLE
Pin memory in Trainer by default

### PR DESCRIPTION
**Pull request by abhishekkrthakur**

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for adding this!

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -244,6 +244,8 @@ class TrainingArguments:
             When using distributed training, the value of the flag :obj:`find_unused_parameters` passed to
             :obj:`DistributedDataParallel`. Will defaut to :obj:`False` if gradient checkpointing is used, :obj:`True`
             otherwise.
+        pin_memory (:obj:`bool`, `optional`, defaults to :obj:`True`)):
+            Whether you want to pin memory in data loaders for PyTorch.
```
Comment (sgugger) [inline src/transformers/training_args.py:248]:
```suggestion
            Whether you want to pin memory in data loaders for PyTorch or not.
```

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -244,6 +244,8 @@ class TrainingArguments:
             When using distributed training, the value of the flag :obj:`find_unused_parameters` passed to
             :obj:`DistributedDataParallel`. Will defaut to :obj:`False` if gradient checkpointing is used, :obj:`True`
             otherwise.
+        pin_memory (:obj:`bool`, `optional`, defaults to :obj:`True`)):
+            Whether you want to pin memory in data loaders for PyTorch or not.
```
Comment (stas00) [inline src/transformers/training_args.py:248]:
```suggestion
            Whether you want to pin memory in data loaders for PyTorch or not.  Will default to  :obj:`True`
```

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -244,6 +244,8 @@ class TrainingArguments:
             When using distributed training, the value of the flag :obj:`find_unused_parameters` passed to
             :obj:`DistributedDataParallel`. Will defaut to :obj:`False` if gradient checkpointing is used, :obj:`True`
```
Comment (stas00) [inline src/transformers/training_args.py:245]:
```suggestion
            :obj:`DistributedDataParallel`. Will default to :obj:`False` if gradient checkpointing is used, :obj:`True`
```

Comment (stas00) [reply src/transformers/training_args.py:245]: this is an old unrelated typo.

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -436,6 +438,7 @@ class TrainingArguments:
             "`DistributedDataParallel`."
         },
     )
+    pin_memory: bool = field(default=True, metadata={"help": "Whether or not to pin memory for PyTorch data loaders."})
```
Comment (stas00) [inline src/transformers/training_args.py:441]:
```suggestion
    pin_memory: bool = field(default=True, metadata={"help": "Whether or not to pin memory for PyTorch DataLoader."})
```

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -242,8 +242,10 @@ class TrainingArguments:
             :obj:`"comet_ml"`, :obj:`"mlflow"`, :obj:`"tensorboard"` and :obj:`"wandb"`.
         ddp_find_unused_parameters (:obj:`bool`, `optional`):
             When using distributed training, the value of the flag :obj:`find_unused_parameters` passed to
-            :obj:`DistributedDataParallel`. Will defaut to :obj:`False` if gradient checkpointing is used, :obj:`True`
+            :obj:`DistributedDataParallel`. Will default to :obj:`False` if gradient checkpointing is used, :obj:`True`
             otherwise.
+        pin_memory (:obj:`bool`, `optional`, defaults to :obj:`True`)):
+            Whether you want to pin memory in data loaders for PyTorch or not.  Will default to :obj:`True`.
```
Comment (stas00) [inline src/transformers/training_args.py:248]:
```suggestion
            Whether you want to pin memory in `DataLoader` or not.  Will default to :obj:`True`.
```

**Comment by stas00** [comment]
Could we please go through normal PR review approval cycles?

It looks like my comment on slack was missed where I suggested to use a more specific cl arg name. 

I proposed:
- dataloader_pin_memory
- dl_pin_memory

But since we already have `dataloader_num_workers`
``
            num_workers=self.args.dataloader_num_workers,
            pin_memory=self.args.pin_memory,
```
it should probably be `dataloader_pin_memory`

This is important since there are other ways to pin memory in pytorch.

**Comment by abhishekkrthakur** [comment]
@stas00 It seems like I missed this message and when I opened this PR in the morning, I didn't see any comments and @sgugger had approved the PR. For a final check, I asked @LysandreJik who gave me the green light.

To avoid this in future, I would request if PR specific comments are made on the PR itself so that author & other reviewers can go through them and make sure that everything is resolved before merging.

**Comment by stas00** [comment]
Yes, absolutely. I guess it just feel through the cracks.

And let's have PR description: as simple as 

This PR adds `--pin_memory` to trainer DataLoader and it defaults to True.

**Pull request merged**